### PR TITLE
Support debugging DMP files

### PR DESCRIPTION
**Pull request by GustavoLCR**
# DO NOT SQUASH

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

* Add Triage Dump (Minidump) parsing support for dmp64 bin plugin
* Support debugging DMP files (Processes, Threads)
* Adds dependency of `rz_type` to `rz_debug` so I can use the type API to get structure offsets from PDB
* Fix some bugs that I found on the way (6cd8f03a8ddd9635a4c27f1fe261f4f881cbf918, fed4eee2bab06cfa9d676b75b0df62b45345dde8, 4ed8f64e90f98008c09efd234d219796a428b1a3, ff56f18fac0d407806f20e56889aec42183f3a47, d8174e84ca05751236cc09d6eaa4d7764cf09404)

Moved from https://github.com/rizinorg/rizin/pull/2171

**Test plan**

* PDB unit tests pass
* DMP tests pass
* WinKD plugin still works

**Closing issues**

Addresses most of #189

### CODE REVIEW

**Comment by XVilka** [comment]
Please also rebase on top of the latest `dev`.
```
 ../librz/bin/format/dmp/dmp64.c: In function ‘rz_bin_dmp64_new_buf’:
../librz/bin/format/dmp/dmp64.c:151:53: error: ‘str.count’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
  151 |                 rz_buf_read(obj->b, (ut8 *)file, str.count * sizeof(ut16));
      |                                                  ~~~^~~~~~
../librz/bin/format/dmp/dmp64.c:139:28: note: ‘str.count’ was declared here
  139 |                 dmp_string str;
      |                            ^~~
../librz/bin/format/dmp/dmp64.c:140:17: error: ‘name_offset’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
  140 |                 rz_buf_seek(obj->b, name_offset, SEEK_SET);
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../librz/bin/format/dmp/dmp64.c:132:22: note: ‘name_offset’ was declared here
  132 |                 ut32 name_offset;
      |                      ^~~~~~~~~~~
```

File: librz/bin/pdb/pdb_downloader.c [hunk 1]
```diff
@@ -23,8 +23,13 @@ static bool download_and_write(SPDBDownloaderOpt *opt, const char *file) {
 		free(dir);
 		return false;
 	}
-	char *url = rz_str_newf("%s/%s/%s/%s", opt->symbol_server, opt->dbg_file, opt->guid, file);
```
Comment (XVilka) [inline librz/bin/pdb/pdb_downloader.c]: 1. Could you please also extract PDB downloading fixes into a separate PR too?
2. Please also rebase on top of the latest `dev`.

**Comment by wargio** [comment]
is this a WIP?

File: librz/debug/p/debug_dmp.c [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_core.h>
+#include <rz_debug.h>
+#include <dmp_specs.h>
+#include <dmp64.h>
+#include <pe_specs.h>
+#include <winkd.h>
+
+// rz_type_db_struct_member_offset() does packed offsets only, we want aligned
```
Comment (XVilka) [inline librz/debug/p/debug_dmp.c]: We should open a general issue to track this and possibly implement in the future.

File: librz/debug/p/debug_dmp.c [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+}
+
+static bool rz_debug_dmp_init(RzDebug *dbg, void **user) {
+	RzCore *core = dbg->corebind.core;
+	RzIODesc *desc = core->io->desc;
+	if (!desc) {
+		return false;
+	}
+	if (strcmp(desc->plugin->name, "dmp")) {
+		eprintf("Open a file with dmp:// to use the 'dmp' debug plugin\n");
```
Comment (XVilka) [inline librz/debug/p/debug_dmp.c]: Please use `RZ_LOG_ERROR()`

File: librz/debug/p/debug_dmp.c [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+		char *pdbpath, *exepath;
+		if (winkd_download_module_and_pdb(&mod, server, symstore, &exepath, &pdbpath)) {
+			dbg->corebind.cmdf(dbg->corebind.core, "idp %s", pdbpath);
+			free(exepath);
+			free(pdbpath);
+			if (!ctx->windctx.profile) {
+				winkd_build_profile(&ctx->windctx, dbg->analysis->typedb);
+			}
+		} else {
+			eprintf("Failed to download ntoskrnl.pdb, many things won't work.\n");
```
Comment (XVilka) [inline librz/debug/p/debug_dmp.c]: Also `RZ_LOG_ERROR()`

File: librz/debug/p/debug_dmp.c [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+		ctx->windctx.target.uniqueid = 0;
+	}
+
+	if (mod.name) {
+		core->bin->cur->o->opts.baseaddr = mod.addr;
+		const char *server = dbg->corebind.cfgGet(dbg->corebind.core, "pdb.server");
+		const char *symstore = dbg->corebind.cfgGet(dbg->corebind.core, "pdb.symstore");
+		char *pdbpath, *exepath;
+		if (winkd_download_module_and_pdb(&mod, server, symstore, &exepath, &pdbpath)) {
+			dbg->corebind.cmdf(dbg->corebind.core, "idp %s", pdbpath);
```
Comment (XVilka) [inline librz/debug/p/debug_dmp.c:127]: Is it possible to do this with the API directly already? Or if not, can be postponed, but ideally, we should avoid calling commands whenever possible.

File: librz/debug/p/debug_dmp.c [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+			if (!ctx->windctx.profile) {
+				winkd_build_profile(&ctx->windctx, dbg->analysis->typedb);
+			}
+		} else {
+			eprintf("Failed to download ntoskrnl.pdb, many things won't work.\n");
+		}
+	}
+
+	if (!ctx->windctx.profile) {
+		eprintf("Could not find a profile for this Windows: %s %" PFMT32d "-bit %" PFMT32u " SP %" PFMT32u "\n",
```
Comment (XVilka) [inline librz/debug/p/debug_dmp.c]: `RZ_LOG_ERROR()`

File: subprojects/rzwinkd/winkd.c [hunk 7]
```diff
@@ -351,231 +361,240 @@ int winkd_walk_vadtree(WindCtx *ctx, ut64 address, ut64 parent) {
...
+	const bool is_target_kernel = ctx->target.uniqueid == 4;
+	if (is_target_kernel) {
+		if (!ctx->PsLoadedModuleList) {
+			eprintf("No PsLoadedModuleList\n");
+			return NULL;
+		}
+		ptr = ctx->PsLoadedModuleList;
+		base = ptr;
+		if (!winkd_read_at_uva(ctx, ptr, (uint8_t *)&ptr, 4 << ctx->is_64bit)) {
+			eprintf("PsLoadedModuleList not present\n");
```
Comment (XVilka) [inline subprojects/rzwinkd/winkd.c]: `RZ_LOG_ERROR()`

File: subprojects/rzwinkd/winkd.c [hunk 8]
```diff
@@ -351,231 +361,240 @@ int winkd_walk_vadtree(WindCtx *ctx, ut64 address, ut64 parent) {
...
+		}
+		if (ptr == base) {
+			return NULL;
+		}
+	} else {
+		ut64 ldroff = ctx->is_64bit ? 0x18 : 0xC;
+		list_entry_off = (4 << ctx->is_64bit) * 2;
+		if (!ctx->target.peb) {
+			eprintf("No PEB\n");
```
Comment (XVilka) [inline subprojects/rzwinkd/winkd.c]: `RZ_LOG_ERROR()`

Comment (XVilka) [reply subprojects/rzwinkd/winkd.c]: Maybe also worth adding a comment that this piece of code does, since there are magical offsets here

File: subprojects/rzwinkd/winkd.c [hunk 9]
```diff
@@ -351,231 +361,240 @@ int winkd_walk_vadtree(WindCtx *ctx, ut64 address, ut64 parent) {
...
+				: 0x44;
 	do {
 
 		ut64 next = 0;
-		winkd_read_at_uva(ctx, (uint8_t *)&next, ptr, 4 << ctx->is_x64);
-		WIND_DBG eprintf("_LDR_DATA_TABLE_ENTRY : 0x%016" PFMT64x "\n", next);
+		winkd_read_at_uva(ctx, ptr, (uint8_t *)&next, 4 << ctx->is_64bit);
 
-		if (!next) {
+		WIND_DBG eprintf("_LDR_DATA_TABLE_ENTRY : 0x%016" PFMT64x "\n", next);
```
Comment (XVilka) [inline subprojects/rzwinkd/winkd.c]: `RZ_LOG_DEBUG()`

File: subprojects/rzwinkd/winkd.c [hunk 10]
```diff
@@ -351,231 +361,240 @@ int winkd_walk_vadtree(WindCtx *ctx, ut64 address, ut64 parent) {
...
+		if (!next || next == UT64_MAX) {
 			eprintf("Corrupted InMemoryOrderModuleList found at: 0x%" PFMT64x "\n", ptr);
```
Comment (XVilka) [inline subprojects/rzwinkd/winkd.c]: `RZ_LOG_ERROR()`

File: subprojects/rzwinkd/winkd.c [hunk 11]
```diff
@@ -351,231 +361,240 @@ int winkd_walk_vadtree(WindCtx *ctx, ut64 address, ut64 parent) {
...
 
-RzList *winkd_list_threads(WindCtx *ctx) {
-	RzList *ret;
-	ut64 ptr, base;
-
-	if (!ctx || !ctx->desc || !ctx->syncd) {
+WindThread *winkd_get_thread_at(WindCtx *ctx, ut64 address) {
+	ut64 entrypoint = 0;
+	if (!ctx->read_at_kernel_virtual(ctx->user, address + O_(ET_Win32StartAddress), (uint8_t *)&entrypoint, 4 << ctx->is_64bit)) {
+		eprintf("Failed to read Win32StartAddress at: 0x%" PFMT64x "\n", address + O_(ET_Win32StartAddress));
```
Comment (XVilka) [inline subprojects/rzwinkd/winkd.c]: `RZ_LOG_ERROR`

File: subprojects/rzwinkd/winkd.c [hunk 12]
```diff
@@ -780,32 +782,28 @@ bool winkd_read_ver(WindCtx *ctx) {
...
+	if (!winkd_read_at(ctx, rr->rz_ver.dbg_addr, (uint8_t *)&ptr, 4 << ctx->windctx.is_64bit)) {
 		free(pkt);
 		return false;
 	}
 
-	ctx->dbg_addr = ptr;
+	ctx->windctx.KdDebuggerDataBlock = ptr;
 
-	WIND_DBG eprintf("_KDDEBUGGER_DATA64 at 0x%016" PFMT64x "\n", ctx->dbg_addr);
+	WIND_DBG eprintf("_KDDEBUGGER_DATA64 at 0x%016" PFMT64x "\n", ctx->windctx.KdDebuggerDataBlock);
```
Comment (XVilka) [inline subprojects/rzwinkd/winkd.c]: `RZ_LOG_DEBUG()`

File: subprojects/rzwinkd/winkd.c [hunk 13]
```diff
@@ -1394,3 +1392,93 @@ int winkd_break_read(WindCtx *ctx) {
 #endif
 	return 1;
 }
+
+static inline char *download_pdb(const char *path, const char *symserver, const char *symstore) {
+	rz_sys_setenv("RZ_BIN_PDBSERVER", symserver);
+	rz_sys_setenv("RZ_BIN_SYMSTORE", symstore);
+	char *direscaped = rz_str_escape_sh(path);
+	char *ret = rz_sys_cmd_strf("rz-bin -PPj %s", direscaped);
```
Comment (XVilka) [inline subprojects/rzwinkd/winkd.c]: Same, is it possible to use the API? cc @ret2libc

**Review by XVilka — CHANGES_REQUESTED** [review]
Amazing job! Looks almost ready to merge, with a few minor things.

File: subprojects/rzwinkd/winkd.c (hunk 13 above)
Comment (ret2libc) [reply subprojects/rzwinkd/winkd.c]: is `rz_bin_pdb_download` the API?

**Comment by GustavoLCR** [comment]
New Changes:
1b539d2 depends on #2229 and
e19d0de depends on  #2230 
Please re-review these commits

File: librz/bin/format/dmp/dmp64.c [hunk 14]
```diff
@@ -68,8 +108,103 @@ static int rz_bin_dmp64_init_header(struct rz_bin_dmp64_obj_t *obj) {
...
+	}
+	ut32 address = obj->triage64_header->DriverListOffset;
 	int i;
+	for (i = 0; i < obj->triage64_header->DriverCount; i++) {
+		dmp_driver_desc *driver = RZ_NEW0(dmp_driver_desc);
+		if (!driver) {
+			break;
+		}
+		ut32 name_offset = 0;
+		const ut64 kldr_entry_addr = address + offsetof(dmp_driver_entry64, LdrEntry);
```
Comment (XVilka) [inline librz/bin/format/dmp/dmp64.c]: Please use `rz_offsetof()`: https://github.com/rizinorg/rizin/blob/dev/librz/include/rz_types.h#L362
```c
// There is a bug of using "offsetof()" in the structure
// initialization in GCC < 5.0 versions
#if !(defined(__GNUC__) && __GNUC__ < 5) || defined(__clang__)
#define rz_offsetof(type, member) offsetof(type, member)
#else
#if __SDB_WINDOWS__
#define rz_offsetof(type, member) ((unsigned long)(ut64) & ((type *)0)->member)
#else
#define rz_offsetof(type, member) ((unsigned long)&((type *)0)->member)
#endif
#endif
```

File: librz/bin/p/bin_dmp64.c [hunk 15]
```diff
@@ -46,19 +46,114 @@ static void header(RzBinFile *bf) {
...
+		rbin->cb_printf("  DataBlocksOffset : 0x%08" PFMT32x "\n", obj->triage64_header->DataBlocksOffset);
+		rbin->cb_printf("  DataBlocksCount : 0x%08" PFMT32x "\n", obj->triage64_header->DataBlocksCount);
 	}
 }
 
+static RzList *fields(RzBinFile *bf) {
+	RzList *fields = rz_list_newf((RzListFree)rz_bin_field_free);
+	struct rz_bin_dmp64_obj_t *obj = (struct rz_bin_dmp64_obj_t *)bf->o->bin_obj;
+#define FIELD_COMMENT(header, field, comment) \
+	rz_list_append(fields, rz_bin_field_new(offsetof(header, field), offsetof(header, field), sizeof(((header *)0)->field), #field, comment, sizeof(((header *)0)->field) == 4 ? "x" : "q", false));
```
Comment (XVilka) [inline librz/bin/p/bin_dmp64.c]: Same, please use `rz_offsetof()`

File: librz/debug/p/debug_dmp.c [hunk 16]
```diff
@@ -0,0 +1,426 @@
...
+
+	ut32 MachineImageType = 0; // Windows Architecture (IMAGE_FILE_MACHINE)
+	ut32 MinorVersion = 0; // Windows Version
+	ut32 ServicePackBuild = 0;
+	ut32 ProcessOffset = 0;
+	ut32 ThreadOffset = 0;
+	ut32 CallStackOffset = 0;
+	ut32 NumberProcessors = 0;
+	RzBuffer *b = rz_buf_new_with_io(&dbg->iob);
+	rz_buf_read_le64_at(b, offsetof(dmp64_header, DirectoryTableBase), &ctx->kernelDirectoryTable);
```
Comment (XVilka) [inline librz/debug/p/debug_dmp.c]: Same, please use `rz_offsetof()`

File: librz/debug/p/debug_dmp.c [hunk 17]
```diff
@@ -0,0 +1,426 @@
...
+	rz_buf_read_le32_at(b, offsetof(dmp64_header, MinorVersion), &MinorVersion);
+	rz_buf_read_le32_at(b, sizeof(dmp64_header) + offsetof(dmp64_triage, ServicePackBuild), &ServicePackBuild);
+	rz_buf_read_le32_at(b, sizeof(dmp64_header) + offsetof(dmp64_triage, ProcessOffset), &ProcessOffset);
+	rz_buf_read_le32_at(b, sizeof(dmp64_header) + offsetof(dmp64_triage, ThreadOffset), &ThreadOffset);
+	rz_buf_read_le32_at(b, sizeof(dmp64_header) + offsetof(dmp64_triage, CallStackOffset), &CallStackOffset);
+	rz_buf_free(b);
+
+	RzIOMap *map = rz_io_map_get(core->io, 0);
+	if (ctx->type == DMP_DUMPTYPE_TRIAGE) {
+		dbg->corebind.cmd(dbg->corebind.core, "e io.va=1");
```
Comment (XVilka) [inline librz/debug/p/debug_dmp.c:64]: Better to use `e io.va=true`. @ret2libc @thestr4ng3r do you think it's possible to use the API somehow? Or it's inevitable for now?

File: librz/io/p/io_dmp.c [hunk 18]
```diff
@@ -0,0 +1,177 @@
+// SPDX-FileCopyrightText: 2021 GustavoLCR <gugulcr@gmail.com>
```
Comment (XVilka) [inline librz/io/p/io_dmp.c]: Shouldn't it be 2022?

File: subprojects/rzwinkd/winkd.c [hunk 19]
```diff
@@ -351,231 +365,244 @@ int winkd_walk_vadtree(WindCtx *ctx, ut64 address, ut64 parent) {
...
+		return NULL;
+	}
+	thread->uniqueid = uniqueid;
+	thread->status = 's';
+	thread->runnable = true;
+	thread->ethread = address;
+	thread->entrypoint = entrypoint;
+	return thread;
+
+	return NULL;
```
Comment (XVilka) [inline subprojects/rzwinkd/winkd.c]: Unnecessary return?

**Review by XVilka — APPROVED** [review]
Please fix that `offsetof()` thing and unnecessary return, and feel free to merge after that, once it's green, @GustavoLCR. The only failure is because of https://github.com/rizinorg/rizin/issues/2313 which is unrelated to this PR.

Amazing work, as always. 💯 👍

File: test/db/formats/dmp/dmp [hunk 20]
```diff
@@ -98,3 +98,249 @@ EXPECT=<<EOF
...
 RUN
+
+NAME=: Triage Dump Debug x64
+FILE=dmp://bins/dmp/triage_x64.dmp
+ARGS=-d
+CMDS=<<EOF
+iH
+e asm.sub.names=false
+pi 4
+dr
```
Comment (yossizap) [inline test/db/formats/dmp/dmp]: Can you also add `dp`, `dpt` and `dpt=` to cover the rest of the plugin's functions?

File: librz/debug/p/debug_dmp.c (hunk 17 above)
Comment (ret2libc) [reply librz/debug/p/debug_dmp.c:64]: Unfortunately I think it is inevitable :(

**Comment by XVilka** [comment]
PDB unit test is failing:
```
 62/101 pdb                            FAIL            0.49s   killed by signal 11 SIGSEGV
>>> MALLOC_PERTURB_=149 /home/runner/work/rizin/rizin/build/test/unit/test_pdb
――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――
stderr:
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
WARNING: rz_type_clone: assertion 'type' failed (line 1169)
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
```

**Comment by XVilka** [comment]
Need to install `cabextract` in CI, I suppose?
```


[XX] db/formats/dmp/dmp : Bitmap Dump Debug
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -d -Qc 'iH
e asm.sub.names=false
pi 4
dr
' dmp://bins/dmp/bmp.dmp
-- stdout
--- expected
+++ actual
@@ -21,26 +21,26 @@
   HeaderSize : 0x0000a038
   BitmapSize : 0x00040000
   Pages : 0x0000544b
-int3
-ret
-int3
-int3
-mxcsr = 0x00001f80
-rflags = 0x00040202
-rax = 0x00000003
-rcx = 0x00000001
-rdx = 0xfffff805135684d0
-rbx = 0xfffff8050f4e9f70
-rsp = 0xfffff805135684f8
-rbp = 0xfffff80513568600
-rsi = 0x00000100
-rdi = 0xfffff8050f4e9f80
-r8 = 0x00000003
-r9 = 0xfffff805135684b8
+invalid
+invalid
+invalid
+invalid
+mxcsr = 0x00000000
+rflags = 0x00000000
+rax = 0x00000000
+rcx = 0x00000000
+rdx = 0x00000000
+rbx = 0x00000000
+rsp = 0x00000000
+rbp = 0x00000000
+rsi = 0x00000000
+rdi = 0x00000000
+r8 = 0x00000000
+r9 = 0x00000000
 r10 = 0x00000000
-r11 = 0xffffa8848825e000
-r12 = 0xfffff8050f4e9f80
-r13 = 0xfffff80510c3c958
+r11 = 0x00000000
+r12 = 0x00000000
+r13 = 0x00000000
 r14 = 0x00000000
-r15 = 0x00000052
-rip = 0xfffff805108776a0
+r15 = 0x00000000
+rip = 0x00000000

-- stderr
WARNING: Failed to read DISPACHER_HEAD.Type at: 0x0
ERROR: NULL value at PsActiveProcessHead

[**]                             db/formats/mach0/fatio    11197 OK       836 BR        0 XX       14 FX


[XX] db/formats/dmp/dmp : Triage Dump Debug x64
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -d -Qc 'iH
e asm.sub.names=false
pi 4
dr
' dmp://bins/dmp/triage_x64.dmp
-- stdout
--- expected
+++ actual
@@ -47,26 +47,26 @@
   DebuggerDataSize : 0x00000380
   DataBlocksOffset : 0x00019438
   DataBlocksCount : 0x00000727
-mov dword [r9 + r8], edx
-jmp 0xfffff8048b58335d
-mov word [r9 + r8], dx
-jmp 0xfffff8048b58335d
-mxcsr = 0x00001f80
-rflags = 0x00050246
-rax = 0xffffc08be4da6240
-rcx = 0x00000020
-rdx = 0x00000001
-rbx = 0x00000001
-rsp = 0xffff850429892120
-rbp = 0xfffff8048b58f598
-rsi = 0xffffc08be457fcf0
-rdi = 0xffffc08be52e7310
-r8 = 0xffff9b0003360008
-r9 = 0x00000044
-r10 = 0x00000020
-r11 = 0x00000020
-r12 = 0x00000015
-r13 = 0xfffff8048b590b01
+invalid
+invalid
+invalid
+invalid
+mxcsr = 0x00000000
+rflags = 0x00000000
+rax = 0x00000000
+rcx = 0x00000000
+rdx = 0x00000000
+rbx = 0x00000000
+rsp = 0x00000000
+rbp = 0x00000000
+rsi = 0x00000000
+rdi = 0x00000000
+r8 = 0x00000000
+r9 = 0x00000000
+r10 = 0x00000000
+r11 = 0x00000000
+r12 = 0x00000000
+r13 = 0x00000000
 r14 = 0x00000000
-r15 = 0x00000015
-rip = 0xfffff8048b58334c
+r15 = 0x00000000
+rip = 0x00000000

-- stderr
sh: 1: cabextract: not found
Couldn't open bin for file '/home/runner/.local/share/rizin/pdb/ntoskrnl.exe/0d8333e61046000/ntoskrnl.exe'
WARNING: Failed to download ntoskrnl.pdb, many things won't work.
ERROR: Could not find a profile for this Windows: x86 64-bit 19041 SP 0
```

**Comment by GustavoLCR** [comment]
> Need to install `cabextract` in CI, I suppose?

@XVilka It's SSL support that's the problem, as we don't build with ssl support enabled by default. I need a way to set the env var `RZ_CURL=1` before rizin launches for the test case to work.

**Comment by XVilka** [comment]
@GustavoLCR shouldn't adding `%RZ_CURL=1` to the test solve your problem?

**Comment by XVilka** [comment]
On FreeBSD:
```
[XX] db/formats/dmp/dmp : Triage Dump Debug x64
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '%RZ_CURL=1
e cfg.debug=1
iH
dL dmp
dpa
dr
s rip
e asm.sub.names=false
pi 4
' dmp://bins/dmp/triage_x64.dmp
-- stdout
--- expected
+++ actual
@@ -48,25 +48,25 @@
   DataBlocksOffset : 0x00019438
   DataBlocksCount : 0x00000727
 mxcsr = 0x00001f80
-rflags = 0x00040286
-rax = 0xffff850429891720
-rcx = 0x0000007e
-rdx = 0xffffffffc0000005
-rbx = 0xfffff8047ba82b68
-rsp = 0xffff850429890ee8
-rbp = 0xffff850429892c10
-rsi = 0x00000000
-rdi = 0x00000000
-r8 = 0xfffff8048b58334c
-r9 = 0xffff850429891ee8
-r10 = 0xffff850429891ee8
-r11 = 0xffff8afd53400000
-r12 = 0xffff850429892c10
-r13 = 0xffff850429891ee8
-r14 = 0xffff850429891560
-r15 = 0xfffff8047ba00000
-rip = 0xfffff8047bdf5a80
-mov qword [rsp + 8], rcx
-mov qword [rsp + 0x10], rdx
-mov qword [rsp + 0x18], r8
-mov qword [rsp + 0x20], r9
+rflags = 0x00050246
+rax = 0xffffc08be4da6240
+rcx = 0x00000020
+rdx = 0x00000001
+rbx = 0x00000001
+rsp = 0xffff850429892120
+rbp = 0xfffff8048b58f598
+rsi = 0xffffc08be457fcf0
+rdi = 0xffffc08be52e7310
+r8 = 0xffff9b0003360008
+r9 = 0x00000044
+r10 = 0x00000020
+r11 = 0x00000020
+r12 = 0x00000015
+r13 = 0xfffff8048b590b01
+r14 = 0x00000000
+r15 = 0x00000015
+rip = 0xfffff8048b58334c
+mov dword [r9 + r8], edx
+jmp 0xfffff8048b58335d
+mov word [r9 + r8], dx
+jmp 0xfffff8048b58335d

-- stderr
Attempting to download compressed pdb in /home/build/.local/share/rizin/pdb/ntoskrnl.exe/0d8333e61046000/ntoskrnl.ex_
Falling back to uncompressed pdb
Attempting to download uncompressed pdb in /home/build/.local/share/rizin/pdb/ntoskrnl.exe/0d8333e61046000/ntoskrnl.exe
ERROR: Cannot open file '/home/build/.local/share/rizin/pdb/ntkrnlmp.pdb/5278AFF86C341677D7D7835C85B7B8441/ntkrnlmp.pdb'
WARNING: Failed to read DISPACHER_HEAD.Type at: 0x0
WARNING: Failed to read DISPACHER_HEAD.Type at: 0x0
WARNING: Failed to read DISPACHER_HEAD.Type at: 0x0
WARNING: Failed to read DISPACHER_HEAD.Type at: 0x0
WARNING: Failed to read DISPACHER_HEAD.Type at: 0x0
WARNING: Failed to read DISPACHER_HEAD.Type at: 0x0
WARNING: Failed to read DISPACHER_HEAD.Type at: 0x0
WARNING: Failed to read DISPACHER_HEAD.Type at: 0x0
WARNING: Failed to read DISPACHER_HEAD.Type at: 0x0
WARNING: Failed to read DISPACHER_HEAD.Type at: 0x0
WARNING: Failed to read DISPACHER_HEAD.Type at: 0x0
WARNING: Failed to read DISPACHER_HEAD.Type at: 0x0
WARNING: Failed to read DISPACHER_HEAD.Type at: 0x0
WARNING: Failed to read DISPACHER_HEAD.Type at: 0x0
WARNING: Failed to read DISPACHER_HEAD.Type at: 0x0
WARNING: Failed to read DISPACHER_HEAD.Type at: 0x0
```

**Review by XVilka — APPROVED** [review]
